### PR TITLE
Fix issues with failed downloads due to SAS-key expiry

### PR DESCRIPTION
**Pull request by ace-e4s**
## Description
This is the final PR addressing the original issue: SAS-key expiry when downloading large/long timeseries. As an added benefit, concurrency is much improved with significant download performance boost!

What has changed:

- Refactor `Storage.get`. Now it accepts a URL (and downloads blobs in sequence and merges). Optimal use is when URL only points to a single day!
- `Storage.get` is much simpler! `BaseDownloader` is gone! Concurrency is moved out to `Client.get`.
- Refactor `Client.get`. It figures out which days are available as data covered by the `start` and `end` range, and makes calls to `Storage.get` accordingly. Multi-threading is applied here for significant perf boost.

As before, unit tests are no good, so test it HARD!

## Checklist
- [x] PR title is descriptive and fit for injection into release notes (see tips below)
- [x] Correct label(s) are used

PR title tips:
* Use imperative mood
* Describe the motivation for change, issue that has been solved or what has been improved - not how
* Examples:
  * Add functionality for Allan variance to sensor_4s.simulate
  * Upgrade to support Python 9.10
  * Remove MacOS from CI

### CODE REVIEW

File: datareservoirio/client.py
Comment (vegard-solum-4ss) [inline datareservoirio/client.py:281]: explotary?

Comment (anna-follestad-4ss) [reply datareservoirio/client.py:281]: exploratory

File: datareservoirio/client.py
Comment (anna-follestad-4ss) [inline datareservoirio/client.py:541]: exploratory

File: datareservoirio/client.py
Comment (anna-follestad-4ss) [inline datareservoirio/client.py:546]: things

File: datareservoirio/storage/storage.py
Comment (anna-follestad-4ss) [inline datareservoirio/storage/storage.py:135]: precedence

File: datareservoirio/storage/storage.py
Comment (anna-follestad-4ss) [inline datareservoirio/storage/storage.py:140]: response

File: tests/test_client.py
Comment (anna-follestad-4ss) [inline tests/test_client.py:308]: L. 308 doesn't seem to do anything?

File: datareservoirio/client.py
Comment (vegard-solum-4ss) [inline datareservoirio/client.py:301]: This will fail when no data is available, since `response_json = {"Files": []}` (which is not `False`).

File: datareservoirio/client.py
Comment (ace-e4s) [inline datareservoirio/client.py:281]:
```suggestion
        # make an "exploratory" request
```

Comment (anna-follestad-4ss) [inline datareservoirio/client.py:281]:
```suggestion
        # make an "exploratory" request
```

File: datareservoirio/client.py
Comment (anna-follestad-4ss) [inline datareservoirio/client.py:541]:
```suggestion
    The list is obtained from the response of an "exploratory" call to
```

File: datareservoirio/client.py
Comment (anna-follestad-4ss) [inline datareservoirio/client.py:546]:
```suggestion
    This is a bit hacky, since it assumes a few things about the response and
```

File: datareservoirio/client.py
Comment (ace-e4s) [reply datareservoirio/client.py:301]: You are right. I'll fix that. Maybe it should be
```
if response_json["Files"]:
```

File: datareservoirio/storage/storage.py
Comment (anna-follestad-4ss) [inline datareservoirio/storage/storage.py:135]:
```suggestion
        The order indicate the priority where item ``n+1`` takes precedence

```

File: datareservoirio/storage/storage.py
Comment (anna-follestad-4ss) [inline datareservoirio/storage/storage.py:140]:
```suggestion
        This is a hacky solution assuming alot about the REST API response from

```

File: tests/test_client.py
Comment (ace-e4s) [reply tests/test_client.py:308]: explain?

File: datareservoirio/client.py
Comment (vegard-solum-4ss) [reply datareservoirio/client.py:301]: Yes, or something like this?
`if days := _timeseries_available_days(response_json)`

File: datareservoirio/storage/storage.py
Comment (anna-follestad-4ss) [inline datareservoirio/storage/storage.py:131]:
```suggestion
        Flatten response from ``/data/days?start={start}&end={end}``

```

File: datareservoirio/storage/storage.py
Comment (vegard-solum-4ss) [inline datareservoirio/storage/storage.py:95]: Any particular reason why you don't use the `get()` method? Just curious.

File: tests/test_client.py
Comment (vegard-solum-4ss) [reply tests/test_client.py:308]: But probably more correct to define it, since it will be called

Comment (anna-follestad-4ss) [reply tests/test_client.py:308]: I tried to give different values (200, 400, 500), test still passed, is that expected behaviour? is it because mock_requests_get.raise_for_status doesn't do anything?

Comment (vegard-solum-4ss) [reply tests/test_client.py:308]: Yes, that is expected since the return value is not used for anything. The `raise_for_status()` method will just be called to see if it raises any exception. But if you give the mock a side_effect where it raises one of these exceptions, the tests will probably fail.

File: datareservoirio/storage/storage.py
Comment (ace-e4s) [reply datareservoirio/storage/storage.py:95]: No, not really. All the `.get`, `.post` etc methods are just thin wrappers around the `.request` method (which it self is a wrapper around `.send` method). 

That said, we should conform to using the high level methods such as `.post` etc instead of `request`, whenever possible.

Comment (ace-e4s) [inline datareservoirio/storage/storage.py:95]:
```suggestion
        response = self._session.get(target_url)

```

File: tests/test_client.py
Comment (ace-e4s) [reply tests/test_client.py:308]: Good discussions. My gut feeling is that its not worth spending time on fixing fundamentally broken unit tests. We need to rewrite the whole shabang anyways.

File: datareservoirio/storage/storage.py
Comment (vegard-solum-4ss) [reply datareservoirio/storage/storage.py:149]: As this code is now, if two or more files contain the same day, only the chunk from the last file will be kept? But as I understand it, the different files can contain parts of the full dataset; therefore, they should all be downloaded and then merged in order to construct the full dataset?

File: datareservoirio/client.py
Comment (anna-follestad-4ss) [inline datareservoirio/client.py:299]: L299 throws a ValueError when trying to get timeseries that are empty

Comment (vegard-solum-4ss) [reply datareservoirio/client.py:299]: Have you pulled the latest changes?

**Review by anna-follestad-4ss — DISMISSED** [review]

**Comment by ace-e4s** [comment]
@anna-follestad-4ss  and/or @vegard-solum-4ss , I will need to ask you to approve again due to small fixed we have done after the approvals. Mostly so that we adhere to the formalities. (I could merge if I wanted to, but...
"Power is when we have every justification to merge... but we don't" 
)

**Comment by ace-e4s** [comment]
One of you is sufficient by the way

**Review by anna-follestad-4ss — APPROVED** [review]

**Pull request merged**
